techlibs/microchip/microchip_dffopt.cc

```
@@ -0,0 +1,340 @@
+					continue;
+				if (cell->getPort(ID(LAT)).is_fully_ones()) // skip latch
+					continue;
+				if (cell->get_bool_attribute(ID::keep)) // keep attribute
+					continue;
+				if (!cell->getPort(ID(ALn)).is_fully_ones()) // async FF
+					continue;
+
+				const bool hasSyncLoad = cell->getPort(ID(SLn)).is_wire();
+				const bool has_s = hasSyncLoad && cell->getPort(ID(SD)).is_fully_ones();
+				const bool has_r = hasSyncLoad && cell->getPort(ID(SD)).is_fully_zero();
+
+				// SLE cannot have both synchronous set and reset implemented at the same time
+				log_assert(!(has_s && has_r));
+
+				// Don't bother if D has more than one use.
+				SigBit sig_D = sigmap(cell->getPort(ID::D));
+				if (bit_uses[sig_D] > 2)
+					continue;
+
+				// Find the D LUT.
+				auto it_D = bit_to_lut.find(sig_D);
+				if (it_D == bit_to_lut.end())
+					continue;
+				LutData lut_d = it_D->second.first;
+				Cell *cell_d = it_D->second.second;
+
+				LutData lut_d_post_ce;
+				LutData lut_d_post_s;
+				LutData lut_d_post_r;
+				bool worthy_post_ce = false;
+				bool worthy_post_s = false;
+				bool worthy_post_r = false;
+
+				// First, unmap CE.
+				SigBit sig_Q = sigmap(cell->getPort(ID::Q));
+				SigBit sig_CE = sigmap(cell->getPort(ID(EN)));
+				LutData lut_ce = LutData(Const(2, 2), {sig_CE}); // INIT = 10
+				auto it_CE = bit_to_lut.find(sig_CE);
+				if (it_CE != bit_to_lut.end())
+					lut_ce = it_CE->second.first;
+				if (sig_CE.wire) {
+					// Merge CE LUT and D LUT into one.  If it cannot be done, nothing to do about this FF.
+					if (!merge_lut(lut_d_post_ce, lut_d, lut_ce, true, sig_Q, max_lut_size))
+						continue;
+
+					// If this gets rid of a CE LUT, it's worth it.  If not, it still may be worth it, if we can remove set/reset
+					// as well.
+					if (it_CE != bit_to_lut.end())
+						worthy_post_ce = true;
+				} else if (sig_CE.data != State::S1) {
+					// Strange.  Should not happen in a reasonable flow, so bail.
```

Maybe a `log_assert` on this to fail noisily if this occurs?